File: web/lib/api/diagnostics.ts
import { apiFetch } from "./client";

export type ComponentStatus = "ok" | "warn" | "err" | "unknown";

export type ComponentId =
  | "llm"
  | "opencode"
  | "env"
  | "worker"
  | "redis"
  | "postgres"
  | "clickhouse";

export interface Metric {
  value: string;
  label: string;
  unit?: string | null;
  suffix?: string | null;
}

export interface DiagnosticComponent {
  id: ComponentId;
  name: string;
  description: string;
  status: ComponentStatus;
  metric?: Metric | null;
  meta: string[];
  message?: string | null;
  causes: string[];
  last_check: string;
  latency_ms?: number | null;
  has_guide: boolean;
}

export interface ConfigGuideStep {
  title: string;
  body: string;
  snippet?: string | null;
}

export interface ConfigGuide {
  component_id: ComponentId;
  title: string;
  intro: string;
  steps: ConfigGuideStep[];
  footer_note?: string | null;
}

export interface DiagnosticsSummary {
  healthy: number;
  warning: number;
  error: number;
  unknown: number;
  total: number;
}

export interface DiagnosticsStatus {
  summary: DiagnosticsSummary;
  last_check: string;
  components: DiagnosticComponent[];
}

export const diagnostics = {
  status: () => apiFetch<DiagnosticsStatus>(`/diagnostics/status`),
  test: (id: ComponentId) =>
    apiFetch<DiagnosticComponent>(`/diagnostics/test/${id}`, { method: "POST" }),
  guide: (id: ComponentId) =>
    apiFetch<ConfigGuide>(`/diagnostics/guide/${id}`),
};

/** Severity rank pra escolher o componente "hero". Maior = mais grave. */
export function severityRank(status: ComponentStatus): number {
  switch (status) {
    case "err":
      return 3;
    case "warn":
      return 2;
    case "unknown":
      return 1;
    case "ok":
    default:
      return 0;
  }
}

/** Formata ISO timestamp como "há Ns" / "há Nmin" / "há Nh". */
export function formatRelativeAge(iso: string): string {
  const target = new Date(iso).getTime();
  const now = Date.now();
  const seconds = Math.max(0, Math.floor((now - target) / 1000));
  if (seconds < 60) return `há ${seconds}s`;
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `há ${minutes} min`;
  const hours = Math.floor(minutes / 60);
  return `há ${hours}h`;
}

File: worker/intelliforce/api/schemas/diagnostics.py
"""Schemas dos endpoints de diagnóstico (tela de Configurações & Saúde).

Diferente de health.py (liveness/readiness pra orquestrador), aqui o foco é
diagnóstico humano: descrição de cada peça, métrica focal, causas prováveis
quando falha.
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


ComponentStatus = Literal["ok", "warn", "err", "unknown"]
ComponentId = Literal[
    "llm", "opencode", "env", "worker", "redis", "postgres", "clickhouse"
]


class Metric(BaseModel):
    """Métrica focal do card (número grande + label mono uppercase)."""
    value: str
    label: str
    unit: str | None = None
    suffix: str | None = None  # ex.: "/13" pra contagens


class DiagnosticComponent(BaseModel):
    """Estado de um componente da plataforma."""
    id: ComponentId
    name: str
    description: str
    status: ComponentStatus
    metric: Metric | None = None
    meta: list[str] = Field(default_factory=list)
    message: str | None = None
    causes: list[str] = Field(default_factory=list)
    last_check: datetime
    latency_ms: float | None = None
    has_guide: bool = False  # se True, frontend mostra botão "Como configurar"


class ConfigGuideStep(BaseModel):
    """Passo numerado dentro de um guia de configuração."""
    title: str
    body: str
    snippet: str | None = None  # bloco de código copiável (env vars, comandos)


class ConfigGuide(BaseModel):
    """Guia humano de como configurar/recuperar um componente."""
    component_id: ComponentId
    title: str
    intro: str
    steps: list[ConfigGuideStep]
    footer_note: str | None = None


class DiagnosticsSummary(BaseModel):
    healthy: int
    warning: int
    error: int
    unknown: int
    total: int


class DiagnosticsStatus(BaseModel):
    """Snapshot completo — saída de GET /diagnostics/status."""
    summary: DiagnosticsSummary
    last_check: datetime
    components: list[DiagnosticComponent]
